README.md
express-respond
=====================

[![Build Status](https://travis-ci.com/lykmapipo/express-respond.svg?branch=master)](https://travis-ci.com/lykmapipo/express-respond)
[![Dependencies Status](https://david-dm.org/lykmapipo/express-respond.svg)](https://david-dm.org/lykmapipo/express-respond)
[![Coverage Status](https://coveralls.io/repos/github/lykmapipo/express-respond/badge.svg?branch=master)](https://coveralls.io/github/lykmapipo/express-respond?branch=master)
[![GitHub License](https://img.shields.io/github/license/lykmapipo/express-respond)](https://github.com/lykmapipo/express-respond/blob/master/LICENSE)

[![Commitizen Friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![Code Style](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)
[![npm version](https://img.shields.io/npm/v/express-respond)](https://www.npmjs.com/package/express-respond)

HTTP response methods with auto content negotiation for [expressjs](https://github.com/strongloop/express/)

## Installation
```sh
$ npm install --save express-respond
```

## Usage
```js
import express from 'express';
import respond from 'express-respond';

const app = express();
app.use(respond);

// respond with 200
app.get('/', (request, response) => {
  response.ok({ name: 'lykmapipo' });
});

// respond with error
app.use((error, request, response, next) => {
  response.error(error);
});
```

## API
All response methods accept optional response `body` based on response type, as some of response are not must to have response body.

### 2xx methods

#### `ok(body: Any)`
Send JSON response with `200 http status code`.

Example
```js
response.ok(body);
```

#### `created(body: Any)`
Send JSON response with `201 http status code`.

Example
```js
response.created(body);
```

#### `accepted([body: Any])`
Send JSON response with `202 http status code`.

Example
```js
response.accepted(body);
```

### `noContent()`
Send JSON response with `204 http status code`.

Example
```js
response.noContent();
```

### 3xx methods

#### `notModified([body: Any])`
Send JSON response with `304 http status code`.

Example
```js
response.notModified(body);
```

### 4xx methods

#### `badRequest([body: Any])`
Send JSON response with `400 http status code`.

Example
```js
response.badRequest(body);
```

#### `unauthorized([body: Any])`
Send JSON response with `401 http status code`.

Example
```js
response.unauthorized(body);
```

#### `paymentRequired([body])`
Send JSON response with `402 http status code`.

Example
```js
response.paymentRequired(body);
```

#### `forbidden([body: Any])`
Send JSON response with `403 http status code`.

Example
```js
response.forbidden(body);
```

#### `notFound([body: Any])`
Send JSON response with `404 http status code`.

Example
```js
response.notFound(body);
```

#### `methodNotAllowed([body: Any])`
Send JSON response with `405 http status code`.

Example
```js
response.methodNotAllowed(body);
```

#### `conflict([body: Any])`
Send JSON response with `409 http status code`.

Example
```js
response.conflict(body);
```

### 5xx methods

#### `internalServerError([body: Any])`
Send JSON response with `500 http status code`.

Example
```js
response.internalServerError(body);
```

#### `notImplemented([body: Any])`
Send JSON response with `501 http status code`.

Example
```js
response.notImplemented(body);
```

#### `badGateway([body: Any])`
Send JSON response with `502 http status code`.

Example
```js
response.badGateway(body);
```

#### `serviceUnavailable([body: Any])`
Send JSON response with `503 http status code`.

Example
```js
response.serviceUnavailable(body);
```

### Error methods

#### `error([body: Error])`
Send `error` as JSON response with `http status code` obtained from `error.status` property.

Example
```js
response.error(error);
```


## Testing

* Clone this repository

* Install all development dependencies

```sh
$ npm install
```
* Then run test

```sh
$ npm test
```

## Contribute

Fork this repo and push in your ideas. 
Do not forget to add a bit of test(s) of what value you adding.

## Licence

The MIT License (MIT)

Copyright (c) lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. 


CHANGELOG.md
#### 1.6.1 (2021-08-23)

##### Chores

* **deps:**  force latest version & audit fix ([4f35514a](https://github.com/lykmapipo/express-respond/commit/4f35514a55e80064c235947f711efd4abf2d19a1))

#### 1.6.0 (2021-08-19)

##### Refactors

*  migrate to es6 ([fcb5f0da](https://github.com/lykmapipo/express-respond/commit/fcb5f0daca3a4690850cd98eb0df3940629a3819))

#### 1.5.15 (2021-02-16)

##### Chores

* **deps:**  force latest version & audit fix ([a055f136](https://github.com/lykmapipo/express-respond/commit/a055f136a1db25dffc1ba88a572aa8db2abfb277))

#### 1.5.14 (2020-12-23)

##### Chores

* **deps:**
  *  force latest version & audit fix ([8dc80510](https://github.com/lykmapipo/express-respond/commit/8dc8051085d4c6ae2b04685f784a018ce220b559))
  *  bump ini from 1.3.5 to 1.3.7 ([756a29a2](https://github.com/lykmapipo/express-respond/commit/756a29a2bdc91b47d2bb4c6228f6af766f696b7d))

#### 1.5.13 (2020-09-13)

##### Chores

* **deps:**  force latest version & audit fix ([48154a61](https://github.com/lykmapipo/express-respond/commit/48154a6147f02f437ed3ef2093cc06c7409003ec))

#### 1.5.12 (2020-09-03)

##### Chores

* **deps:**  force latest version & audit fix ([8b365333](https://github.com/lykmapipo/express-respond/commit/8b36533380db840a7824647bd14b58272e11916a))

#### 1.5.11 (2020-08-06)

##### Chores

* **deps:**  force latest version & audit fix ([b332c0b9](https://github.com/lykmapipo/express-respond/commit/b332c0b92056f519f0fffab2ba085e15d6e097af))

#### 1.5.10 (2020-07-28)

##### Chores

* **deps:**  force latest version & audit fix ([034a53d3](https://github.com/lykmapipo/express-respond/commit/034a53d34effc4d7c521f9e2b9084ab9e7e8a4ff))

#### 1.5.9 (2020-07-21)

##### Chores

* **deps:**  force latest version & audit fix ([05466ca2](https://github.com/lykmapipo/express-respond/commit/05466ca24dd44896ba85fdb2e5cc94d80a6baaf2))

#### 1.5.8 (2020-07-12)

##### Chores

* **deps:**  force latest version & audit fix ([3f664666](https://github.com/lykmapipo/express-respond/commit/3f6646664606693a991c520f5c0163e065a73e03))

#### 1.5.7 (2020-07-02)

##### Chores

* **deps:**  force latest version & audit fix ([fe03e2c4](https://github.com/lykmapipo/express-respond/commit/fe03e2c414d138a8101c9a3e4c2422618c5648af))

#### 1.5.6 (2020-06-18)

##### Chores

* **deps:**  force latest version & audit fix ([2fd263bf](https://github.com/lykmapipo/express-respond/commit/2fd263bfb99211bcafb3d245af7b9735c7a1579c))

#### 1.5.5 (2020-06-09)

##### Chores

* **deps:**  force latest version & audit fix ([538cfa23](https://github.com/lykmapipo/express-respond/commit/538cfa23a30e2f9d86f7577f53b84cad09b7ccd1))

#### 1.5.4 (2020-05-31)

##### Chores

* **deps:**  force latest version & audit fix ([b2e70b79](https://github.com/lykmapipo/express-respond/commit/b2e70b79027c35f6b46c2626adb78797d0f1d21e))

#### 1.5.3 (2020-05-19)

##### Chores

* **deps:**  force latest version & audit fix ([37a9e5e5](https://github.com/lykmapipo/express-respond/commit/37a9e5e5f9700dc26e721944e780fdff4955dc59))

#### 1.5.2 (2020-05-10)

##### Chores

* **deps:**  force latest version & audit fix ([d8e7c593](https://github.com/lykmapipo/express-respond/commit/d8e7c59378b9773e0e2615e7d5ec1fd061c17625))

#### 1.5.1 (2020-04-24)

#### 1.5.0 (2020-04-24)

##### Refactors

*  use latest http statuses ([9678b686](https://github.com/lykmapipo/express-respond/commit/9678b686835f4e79c6e4f3c21d2558494990f0d0))

#### 1.4.23 (2020-04-18)

##### Chores

* **deps:**  force latest version & audit fix ([b826eb17](https://github.com/lykmapipo/express-respond/commit/b826eb17ed557f75cddd2449d4e6cd56f65168f3))

#### 1.4.22 (2020-03-30)

#### 1.4.21 (2020-03-06)

##### Chores

* **deps:**  force latest version & audit fix ([fda2ed58](https://github.com/lykmapipo/express-respond/commit/fda2ed58245ee68b04ffac4094ac18e8091446e3))

#### 1.4.20 (2020-02-28)

##### Chores

* **deps:**  force latest version & audit fix ([c120aed7](https://github.com/lykmapipo/express-respond/commit/c120aed72083219fba7dc5300492eecd592df961))

#### 1.4.19 (2020-02-16)

##### Chores

* **deps:**  force latest version & audit fix ([f0d07a58](https://github.com/lykmapipo/express-respond/commit/f0d07a58457441731af96c9fbfacbd985173490b))

#### 1.4.18 (2020-02-04)

##### Chores

* **deps:**  force latest version & audit fix ([498a25a0](https://github.com/lykmapipo/express-respond/commit/498a25a0a4d4d7e5fb39e796f66546e3a87bb84f))

#### 1.4.17 (2020-01-22)

##### Chores

* **deps:**  force latest version & audit fix ([71e52a9a](https://github.com/lykmapipo/express-respond/commit/71e52a9a5aafab35dd844337096784ed5fd15279))

#### 1.4.16 (2020-01-17)

##### Chores

* **deps:**  force latest version & audit fix ([9df4e13f](https://github.com/lykmapipo/express-respond/commit/9df4e13f0d8da3ae9399eec4fa7f2571919fb8c4))

#### 1.4.15 (2020-01-12)

##### Chores

* **deps:**  force latest version & audit fix ([c081b716](https://github.com/lykmapipo/express-respond/commit/c081b716eb2f1c33c76df664ee0885f032b83144))

#### 1.4.14 (2019-12-18)

##### Chores

* **deps:**  force latest version & audit fix ([350d740c](https://github.com/lykmapipo/express-respond/commit/350d740cf8437bfbf85c6b0751b3779ee7df2c6e))

#### 1.4.13 (2019-11-29)

##### Chores

* **deps:**  force latest version & audit fix ([46288656](https://github.com/lykmapipo/express-respond/commit/46288656c6bb54c36f9d4b4b34578d8244307955))

#### 1.4.12 (2019-10-23)

##### Chores

* **deps:**  force latest version & audit fix ([5b7b92f8](https://github.com/lykmapipo/express-respond/commit/5b7b92f8ac60dff8e78be17838a6fffc79c3df75))

##### Refactors

*  remove include usage ([698085f1](https://github.com/lykmapipo/express-respond/commit/698085f17c739a8f17a635bfb77a8f6b204640d5))

#### 1.4.11 (2019-10-10)

##### Chores

* **deps:**  force latest version & audit fix ([4e392d72](https://github.com/lykmapipo/express-respond/commit/4e392d72ee11d6f00bf50a46b24eed9153f6876b))

#### 1.4.10 (2019-09-24)

##### Chores


lib/index.js
'use strict';

const _ = require('lodash');
const common = require('@lykmapipo/common');
const env = require('@lykmapipo/env');

// TODO: just extend response directly

/**
 * @function prepareBody
 * @name prepareBody
 * @description prepare response body
 * @param {*} [data] valid response body
 * @param {string} [code=500] valid http response status code
 * @returns {object} formatted response body
 * @author lally elias <lallyelias87@mail.com>
 * @since  1.1.0
 * @version 0.1.0
 * @license MIT
 * @private
 * @example
 * const body = prepareBody(new Error('Missing API Key'));
 * //=> { message: 'Missing API Key', ... }
 *
 * const body = prepareBody({ name: 'lykmapipo' });
 * //=> { name: 'lykmapipo' }
 */
const prepareBody = (data, code = 500) => {
  // build response base on http status code
  if (!data) {
    const status = code;
    const message = common.STATUS_CODES[code];
    const description = common.STATUS_CODES[code];
    return { status, code, message, description };
  }

  // handle error or data
  const stack = !env.isProduction();
  let body = {};
  body = _.isError(data) ? common.mapErrorToObject(data, { code, stack }) : data;

  // return formatted response body
  return body;
};

/**
 * @function mapToHttpReply
 * @name mapToHttpReply
 * @description generate http reply method based on http status
 * @param {object} response valid express http response object
 * @param {string} code valid http response status code
 * @returns {Function} a valid response function helper
 * @author lally elias <lallyelias87@mail.com>
 * @since  1.2.0
 * @version 0.1.0
 * @license MIT
 * @private
 */
const mapToHttpReply = (response, code) => (data) => {
  // prepare http status
  let status = _.get(data, 'status', code);

  // FIX: RangeError [ERR_HTTP_INVALID_STATUS_CODE]: Invalid status code
  // FIX: if response data has status field for business logics
  status = _.isNumber(status) || _.isString(status) ? status : code;

  // set response status
  response.status(status);

  // respond with json body
  const ignoreBody = _.includes([204, 304], _.toNumber(status));
  if (ignoreBody) {
    return response.json();
  }

  // prepare response body
  const body = prepareBody(data, status);
  return response.json(body);
};

/**
 * @function mapStatusToMethod
 * @name mapStatusToMethod
 * @description map http status to named response method
 * @param {object} response valid express http response object
 * @returns {Function} a valid response function helper
 * @author lally elias <lallyelias87@mail.com>
 * @since  1.2.0
 * @version 0.1.0
 * @license MIT
 * @private
 */
const mapStatusToMethod = (response) => (status, code) => {
  // prepare response method from http status message
  const method = _.camelCase(status);

  // extend http response with the custom response type method
  const httpReply = mapToHttpReply(response, code);
  response[code] = httpReply;
  response[method] = httpReply;
};

/**
 * @function respond
 * @name respond
 * @description common http responses for expressjs
 * @param {object} request valid express http request object
 * @param {object} response valid express http response object
 * @param {Function} next a middleware to invoke for continuation
 * @returns {Function} a valid expressjs middleware
 * @author lally elias <lallyelias87@mail.com>
 * @since  1.0.0
 * @version 0.1.0
 * @license MIT
 * @public
 * @example
 * const express = require('express');
 * const respond = require('express-respond');
 *
 * const app = express();
 * app.use(respond);
 *
 * app.get('/', (request, response) => {
 *   response.ok({ name: 'lykmapipo' });
 * });
 */
const respond = (request, response, next) => {
  // map http status code to http response method
  _.forEach(common.STATUS_CODES, mapStatusToMethod(response));

  // add error response method
  response.error = mapToHttpReply(response, 500);

  // continue with middleware chain
  return next();
};

module.exports = respond;
